docs/specs/user-profile/design.md

```diff
@@ -0,0 +1,329 @@
+> We only support `YYYY-MM-DD` for simplicity.
+
+### given\_name
+
+The given name of the first name of the end-user.
+The input control is a simple text input without validation.
+It is subject to [automatic population](#automatic-population).
+
+### family\_name
+
+The family name or the last name of the end-user.
+The input control is a simple text input without validation.
+It is subject to [automatic population](#automatic-population).
+
+## Automatic population
+
+> Automatic population CANNOT be turned off!
+
+Some of the standard attributes are subject to automatic population.
+
+The automatic population runs when one of the following situation happens:
+
+- When the end-user signs up.
+- When the end-user adds an identity.
+- When the end-user updates an identity.
+- When the end-user removes an identity.
+
+The steps of the automatic population are:
+
+1. Generate a list of email address candidates from the identities, candidates from newer identities are ordered first.
+1. Generate a list of phone number candidates from the identities, candidates from newer identities are ordered first.
+1. Generate a list of username candidates from the identities, candidates from newer identities are ordered first.
+1. Generate a list of given name and family name candidates from the identities, candidates from newer identities are ordered first.
+1. If the `email` standard attribute does not refer to a candidate in the list, clear it.
+1. If the `phone_number` standard attribute does not refer to a candidate in the list, clear it.
+1. If the `preferred_username` standard attribute does not refer to a candidate in the list, clear it.
+1. If the `email` standard attribute is absent and the list of candidate is non-empty, set it to the first candidate in the list.
+1. If the `phone_number` standard attribute is absent and the list of candidate is non-empty, set it to the first candidate in the list.
+1. If the `preferred_username` standard attribute is absent and the list of candidate is non-empty, set it to the first candidate in the list.
+1. If both the `given_name` and the `family_name` standard attribute are absent and the list of candidate is non-empty, set it to the first candidate in the list.
+
+## Custom attributes
+
+In addition to standard attributes, the developer can define their own custom attributes.
+
+### Defining custom attributes
+
+The custom attributes are defined as a single JSON schema written against a subset of JSON schema 2019-09.
+
+Here is an example of the schema of the custom attributes.
+
+```JSON
+{
+  "properties": {
+    "app_user_role": {
+      "type": "string",
+      "enum": ["owner", "editor", "viewer"]
+    },
+    "stripe_customer_id": {
+      "type": "string"
+    }
+  }
+}
+```
+
+All changes made to custom attributes must be valid against the schema.
+
+#### Supported subset of JSON schema 2019-09
+
+- `type`: `boolean`, `string`, `number`, `integer`
```

Comment:
I think we need way more than these types, at least if we look at the standard attributes, there are more types like `email`, `phone`, `url`, `photo`, `date`, `time`

Reply:
JSON schema has a keyword called format.
We can define email, phone, url, and datetime as format.
email can just be rendered using a text input.
phone can be rendered with the existing library we are using.
url can just be rendered using a text input, with validation to ensure the value is an URL.
If there is an existing library for date, time, datetime, then we can support them.